app/models/manageiq/providers/redhat/infra_manager.rb

```diff
@@ -103,24 +103,39 @@ def vm_reconfigure(vm, options = {})
   end
 
   def add_disks(add_disks_spec, vm)
-    ems_storage_uid = add_disks_spec["ems_storage_uid"]
+    storage = add_disks_spec["storage"]
     with_disk_attachments_service(vm) do |service|
-      add_disks_spec["disks"].each { |disk_spec| service.add(prepare_disk(disk_spec, ems_storage_uid)) }
+      add_disks_spec["disks"].each { |disk_spec| service.add(prepare_disk(disk_spec, storage)) }
     end
   end
 
-  def prepare_disk(disk_spec, ems_storage_uid)
-    {
-      :bootable  => disk_spec["bootable"],
-      :interface => "VIRTIO",
-      :active    => true,
-      :disk      => {
-        :provisioned_size => disk_spec["disk_size_in_mb"].to_i * 1024 * 1024,
-        :sparse           => disk_spec["thin_provisioned"],
-        :format           => disk_spec["format"],
-        :storage_domains  => [:id => ems_storage_uid]
-      }
-    }
+  # prepare disk attachment request payload of adding disk for reconfigure vm
+  def prepare_disk(disk_spec, storage)
+    disk_attachment_builder = DiskAttachmentBuilder.new
+    disk_attachment_builder.storage = storage
+    disk_attachment_builder.thin_provisioned = disk_spec[:thinProvisioned]
+    disk_attachment_builder.bootable = disk_spec["bootable"]
+    disk_attachment_builder.name = disk_spec[:diskName]
+    disk_attachment_builder.size_in_mb = disk_spec["disk_size_in_mb"]
+    disk_attachment_builder.disk_attachment
+  end
+
+  # add disk to a virtual machine for a request arrived from an automation call
+  def vm_add_disk(vm, options = {})
+    storage = options[:datastore] || vm.storage
+    raise _("Data Store does not exist, unable to add disk") unless storage
+
+    disk_attachment_builder = DiskAttachmentBuilder.new
```

Comment:
I think you should just pass options to the attachment builder (in the initializer) and encapsulate dealing with them there